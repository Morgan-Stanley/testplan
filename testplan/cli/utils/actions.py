from typing import Iterable

from testplan.report import TestReport


class ParseSingleAction:
    def __call__(self) -> TestReport:
        pass


class ParseMultipleAction:
    def __call__(self) -> Iterable[TestReport]:
        pass


class ProcessResultAction:
    def __call__(self, result: TestReport) -> TestReport:
        pass
