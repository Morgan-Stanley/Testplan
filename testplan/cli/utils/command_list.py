import click


class CommandList:
    def __init__(self):
        self.commands = []

    def command(self, *args, **kwargs):
        def inner(func):
            command = click.command(*args, **kwargs)(func)
            self.commands.append(command)
            return command

        return inner

    def register_to(self, group: click.Group) -> None:
        for command in self.commands:
            group.add_command(command)
