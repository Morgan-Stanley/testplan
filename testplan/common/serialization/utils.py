import warnings
from typing import Any

import dill

FIFTY_MEGA = 52428800


class TestplanSerializationError(Exception):
    """general serialization error"""


class TestplanDeserializationError(Exception):
    """general deserialization error"""


def serialize(obj: Any) -> bytes:
    try:
        data = dill.dumps(obj, byref=True)
        if len(data) > FIFTY_MEGA:
            warnings.warn(f"Too big object {obj} after serialization.")
        return data
    except (TypeError, dill.PickleError) as exc:
        raise TestplanSerializationError(str(obj)) from exc


def deserialize(data: bytes) -> Any:
    try:
        return dill.loads(data)
    except EOFError:
        raise TestplanDeserializationError(
            "No data input for deserialization."
        )
    except (TypeError, dill.PickleError) as exc:
        raise TestplanDeserializationError() from exc
