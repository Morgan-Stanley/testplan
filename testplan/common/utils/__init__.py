"""Common utility modules."""

from contextlib import contextmanager


# TODO: typing, tests

@contextmanager
def compose_contexts(*args):
    """
    Roll context managers together, entering from left to right, exiting from right to left.
    """

    if not args:
        raise TypeError("compose_contexts: have nothing to compose")
    if len(args) == 1:
        with args[0] as a:
            yield a
    else:
        head, tail = args[0], args[1:]
        with head as a:
            with compose_contexts(*tail) as b:
                if isinstance(b, tuple):
                    yield (a, *b)
                else:
                    yield (a, b)
