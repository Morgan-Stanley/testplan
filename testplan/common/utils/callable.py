"""Utilities related to python callables (functions, methods, classes etc.)"""

import inspect
import functools
from collections import namedtuple
from collections.abc import Sequence

WRAPPER_ASSIGNMENTS = functools.WRAPPER_ASSIGNMENTS + (
    "__tags__",
    "__tags_index__",
    "summarize",
    "summarize_num_passing",
    "summarize_num_failing",
)

# Local copy of inspect.ArgSpec namedtuple - see notes on inspect.getargspec()
# deprecation within getargspec() below.
ArgSpec = namedtuple("ArgSpec", ["args", "varargs", "keywords", "defaults"])


def arity(function):
    """
    Return the arity of a function

    :param function: function
    :type function: ``function``

    :return: arity of the function
    :rtype: ``int``
    """
    return len(getargspec(function).args)


def getargspec(callable_):
    """
    Return an Argspec for any callable object

    :param callable_: a callable object
    :type callable_: ``Callable``

    :return: argspec for the callable
    :rtype: ``ArgSpec``
    """
    if not callable(callable_):
        raise ValueError("{} is not callable".format(callable_))

    if inspect.ismethod(callable_) or inspect.isfunction(callable_):
        func = callable_
    else:
        func = callable_.__call__

    # In Python 3.7 inspect.getargspec() is deprecated and will be removed in
    # 3.8, due to the addition of keyword-only args and type annotations
    # (see PEPs 3102 and 484 for more information). To retain backwards
    # compatibility we convert from a FullArgSpec to a python2 ArgSpec.
    full_argspec = inspect.getfullargspec(func)

    # Raise a ValueError if the function has any keyword-only args defined,
    # since we can't easily handle them in a way that is also python 2
    # compatible. On the other hand, we can just discard any information on
    # type annotations.
    if full_argspec.kwonlyargs:
        raise ValueError(
            "Cannot get argspec for function with keyword-only args "
            "defined: {}".format(func)
        )
    return ArgSpec(
        args=full_argspec.args,
        varargs=full_argspec.varargs,
        keywords=full_argspec.varkw,
        defaults=full_argspec.defaults,
    )


def wraps(
    wrapped, assigned=WRAPPER_ASSIGNMENTS, updated=functools.WRAPPER_UPDATES
):
    """
    Custom wraps function that copies some additional attr than default.
    """

    def _inner(wrapper):
        wrapper = functools.update_wrapper(
            wrapper=wrapper,
            wrapped=wrapped,
            assigned=assigned,
            updated=updated,
        )
        return wrapper

    return _inner


def pre(*prefunctions):
    """
    Attaches function(s) to another function for systematic execution before
    said function, with the same arguments

    :param prefunction: function to execute before
    :type prefunction: ``callable``

    :return: function decorator
    :rtype: ``callable``
    """

    def outer(function):
        """
        Function decorator adding the execution of a prefunction
        """

        @wraps(function)
        def inner(*args, **kwargs):
            """
            Inner decorator body, executing the prefunctions and the
            function itself
            """
            for prefunction in prefunctions:
                prefunction(*args, **kwargs)
            return function(*args, **kwargs)

        return inner

    return outer


def post(*postfunctions):
    """
    Attaches function(s) to another function for systematic execution after
    said function, with the same arguments

    :param postfunction: function to execute after
    :type postfunction: ``callable``

    :return: function decorator
    :rtype: ``callable``
    """

    def outer(function):
        """
        Function decorator adding the excution of a postfunction
        """

        @wraps(function)
        def inner(*args, **kwargs):
            """
            Inner decorator body, executing the postfunctions and the
            function itself
            """
            result = function(*args, **kwargs)
            for postfunction in postfunctions:
                postfunction(*args, **kwargs)
            return result

        return inner

    return outer


def functiondispatch(func, in_list=False):
    """
    Like `singledispatch` but dispatches by type of either the first
    argument, or the first element of that argument which is a sequence.
    """
    registry = {}

    def dispatch(*args):
        if not args:
            raise TypeError(
                f"{getattr(func, '__name__', 'dispatch function')}"
                " requires at least 1 positional argument"
            )
        if in_list and issubclass(args[0].__class__, Sequence) and args[0][0]:
            cls = args[0][0].__class__
        else:
            cls = args[0].__class__
        return registry[cls] if cls in registry else func

    def register(cls):
        def wrap(func):
            if cls in registry:
                raise ValueError(
                    "There is already a handler registered"
                    f" for {cls!r} in `@type_dispatch`"
                )
            registry[cls] = func
            return func

        return wrap

    @functools.wraps(func)
    def wrapper(*args, **kwargs):
        return dispatch(*args)(*args, **kwargs)

    wrapper.register = register
    wrapper.dispatch = dispatch
    functools.update_wrapper(wrapper, func)
    return wrapper


def dispatchmethod(in_list=False):
    """
    Like `singledispatchmethod` but dispatches by type of either the first
    argument, or the first element of that argument which is a sequence.

    Supports wrapping existing descriptors and handles non-descriptor callables
    as instance methods.
    """

    class DispatchMethod(object):
        """`functiondispatch` generic method descriptor."""

        def __init__(self, func):
            if not callable(func) and not hasattr(func, "__get__"):
                raise TypeError(f"{func!r} is not callable or a descriptor")

            self.dispatcher = functiondispatch(func, in_list=in_list is True)
            self.func = func

        def register(self, cls):
            """
            Register a new implementation for the given `cls` on
            a generic method.
            """
            return self.dispatcher.register(cls)

        def __get__(self, obj, cls=None):
            def _method(*args, **kwargs):
                method = self.dispatcher.dispatch(*args)
                return method.__get__(obj, cls)(*args, **kwargs)

            _method.__isabstractmethod__ = self.__isabstractmethod__
            _method.register = self.register
            functools.update_wrapper(_method, self.func)
            return _method

        @property
        def __isabstractmethod__(self):
            return getattr(self.func, "__isabstractmethod__", False)

    if callable(in_list):
        # decorator is used without any argument
        return DispatchMethod(in_list)
    else:
        # argument `in_seq` is explicitly specified
        return DispatchMethod
