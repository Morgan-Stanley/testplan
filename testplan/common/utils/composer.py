"""
Place where we put a variety of higher-order stuffs.
"""


from types import TracebackType
from typing import ContextManager, Type


class ComposedContextManager:
    """
    Roll context managers together,
    entering from left to right,
    exiting from right to left.

    The return value of __exit__ should only matter when exception exists.
    If __exit__ return some truthy value, we know this current exception is
    caught.
    """

    def __init__(self, head: ContextManager, *tail: ContextManager):
        self.head = head
        self.tail = None
        if len(tail):
            self.tail = ComposedContextManager(*tail)

    def __enter__(self):
        r = self.head.__enter__()
        if self.tail is not None:
            rs = self.tail.__enter__()
            if isinstance(rs, tuple):
                return (r, *rs)
            else:
                return (r, rs)
        else:
            return r

    def __exit__(
        self,
        exc_type: Type[BaseException],
        exc_value: BaseException,
        traceback: TracebackType,
    ):
        if self.tail is not None:
            r = self.tail.__exit__(exc_type, exc_value, traceback)
            # If r is a truthy value, the exception is handled.
            if exc_type and not r:
                return self.head.__exit__(exc_type, exc_value, traceback)
            # We know self.tail.__exit__ has already handled the exception,
            # it's just the parameters are still not None here.
            self.head.__exit__(None, None, None)
            return True
        else:
            return self.head.__exit__(exc_type, exc_value, traceback)


def compose_contexts(
    head: ContextManager, *tail: ContextManager
) -> ContextManager:
    return ComposedContextManager(head, *tail)
