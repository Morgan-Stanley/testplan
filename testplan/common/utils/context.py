"""TODO."""
import warnings
from typing import Union, Dict

from testplan.vendor.tempita import Template as TempitaTemplate
from jinja2 import Template


def parse_template(template: str) -> Union[TempitaTemplate, Template]:
    tempita_failed = False
    parsed_template = None
    try:
        parsed_template = Template(template)
    except Exception:
        try:
            # Jinja failed try with tempita
            parsed_template = TempitaTemplate(template)
        except Exception:
            tempita_failed = True
        else:
            tempita_warning = f"The template: '{template}' is not a valid Jinja2 template. Falling back to Tempita. Tempita will be decommisioned soon, please update your template."
            warnings.warn(tempita_warning, FutureWarning)
            pass
        finally:
            if tempita_failed:
                # raise the original jinja exception so user will fix it for jinja
                raise

    return parsed_template


class ContextValue:
    """
    A context value represents a combination of a driver name
    and a tempita template, to be resolved on driver start.
    """

    def __init__(self, driver: str, value: str):
        """
        Create a new context value
        """
        self.driver = driver
        self.template = parse_template(value)

    def __call__(self, ctx):
        """
        Resolve the template.
        """
        if ctx is None:
            raise ValueError(
                f'Could not retrieve driver "{self.driver}" value'
                " from NoneType context."
            )
        if self.driver not in ctx:
            raise RuntimeError(
                f'Driver "{self.driver}" is not present in context.'
            )
        return render(self.template, ctx[self.driver].context_input())


def context(driver, value):
    """
    Create a context extractor from a driver name and a value
    expression. Value expressions must be valid tempita templates,
    which will be resolved from the context.
    """
    return ContextValue(driver, value)


def is_context(value):
    """
    Checks if a value is a context value
    :param value: Value which may have been constructed through `context`
    :type value: ``object``

    :return: True if it is a context value, otherwise False
    :rtype: ``bool``
    """
    return isinstance(value, ContextValue)


def expand(value, contextobj, constructor=None):
    """
    Take a value and a context and return the expanded result.
    Apply a constructor if necessary.
    """
    if is_context(value):
        expanded_value = value(contextobj)
        if constructor:
            expanded_value = constructor(expanded_value)
        return expanded_value
    else:
        return value


def expand_env(
    orig: Dict[str, str],
    overrides: Dict[str, Union[str, ContextValue]],
    contextobj,
):
    """
    Copies the `orig` dict of environment variables.
    Applies specified overrides.
    Removes keys that have value of None as override.
    Expands context values as strings.
    Returns as a copy.

    :param orig: The initial environment variables. Usually `os.environ` is to
                 be passed in. This will not be modified.
    :type orig: ``dict`` of ``str`` to ``str``
    :param overrides: Keys and values to be overriden. Values can be strings or
                      context objects.
    :type overrides: ``dict`` of ``str`` to either ``str`` or ``ContextValue``
    :param contextobj: The context object that can be used to expand context
                       values.
    :type contextobj: ``object``

    :return: Copied, overridden and expanded environment variables
    :rtype: ``dict``
    """
    env = orig.copy()
    env.update(overrides)
    return {
        key: expand(val, contextobj, str)
        for key, val in env.items()
        if val is not None
    }


def render(template: Union[Template, TempitaTemplate, str], context):
    if isinstance(template, str):
        template = parse_template(template)

    return (
        template.substitute(context)
        if isinstance(template, TempitaTemplate)
        else template.render(context)
    )
