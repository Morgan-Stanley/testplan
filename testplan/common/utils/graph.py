from collections import defaultdict
from copy import copy
from dataclasses import dataclass
from typing import Dict, Generic, List, Tuple, TypeVar

try:
    from typing import Self  # starting from py311
except ImportError:
    from typing_extensions import Self

from memoization import cached

T = TypeVar("T")  # vertex rep
U = TypeVar("U")  # actual vertex
V = TypeVar("V")  # actual edge

GraphComponent = List[T]  # a component is a group of vertices


@dataclass
class DirectedGraph(Generic[T, U, V]):
    """adjacency matrix for directed graph"""

    vertices: Dict[T, U]
    edges: Dict[T, Dict[T, V]]
    indegrees: Dict[T, int]
    outdegrees: Dict[T, int]

    @classmethod
    def new(cls) -> Self:
        return cls(dict(), dict(), dict(), dict())

    @classmethod
    def from_vertices(cls, vertices: Dict[T, U]) -> Self:
        return cls(
            copy(vertices),
            {k: dict() for k in vertices},
            {k: 0 for k in vertices},
            {k: 0 for k in vertices},
        )

    @classmethod
    def from_vertices_and_edges(
        cls, vertices: Dict[T, U], edges: Dict[Tuple[T, T], V]
    ) -> Self:
        edges_ = {k: dict() for k in vertices}
        indegrees_ = {k: 0 for k in vertices}
        outdegrees_ = {k: 0 for k in vertices}
        for (src, dst), val in edges.items():
            edges_[src][dst] = val
            indegrees_[dst] += 1
            outdegrees_[src] += 1

        return cls(copy(vertices), edges_, indegrees_, outdegrees_)

    def __invalid(self):
        self.cycles.cache_clear()

    def add_vertex(self, rep: T, obj: U) -> bool:
        if rep in self.vertices:
            return False
        self.__invalid()
        self.vertices[rep] = obj
        self.edges[rep] = dict()
        self.indegrees[rep] = 0
        self.outdegrees[rep] = 0
        return True

    def add_edge(self, src: T, dst: T, edge_val: V) -> bool:
        if (
            src not in self.edges
            or dst not in self.edges
            or dst in self.edges[src]
        ):
            return False
        self.__invalid()
        self.edges[src][dst] = edge_val
        self.indegrees[dst] += 1
        self.outdegrees[src] += 1
        return True

    def remove_vertex(self, rep: T) -> bool:
        if rep not in self.vertices:
            return False

        self.__invalid()
        del self.vertices[rep]
        del self.indegrees[rep]
        del self.outdegrees[rep]

        for v in self.edges[rep]:
            self.indegrees[v] -= 1
        del self.edges[rep]

        for v in self.vertices:
            if rep in self.edges[v]:
                del self.edges[v][rep]
                self.outdegrees[v] -= 1
        return True

    def remove_edge(self, src: T, dst: T) -> bool:
        if (
            src not in self.edges
            or dst not in self.edges
            or dst not in self.edges[src]
        ):
            return False

        self.__invalid()
        del self.edges[src][dst]
        self.outdegrees[src] -= 1
        self.indegrees[dst] -= 1
        return True

    def update_vertex(self, rep: T, new_obj: U) -> bool:
        if not rep in self.vertices:
            return False
        self.vertices[rep] = new_obj
        return True

    def update_edge(self, src: T, dst: T, new_val: V) -> bool:
        if (
            src not in self.edges
            or dst not in self.edges
            or dst not in self.edges[src]
        ):
            return False
        self.edges[src][dst] = new_val
        return True

    def tarjan_scc(self) -> Dict[T, GraphComponent[T]]:
        """Tarjan's strongly connected components algorithm"""
        index = 0
        stack = []
        indices = dict(map(lambda x: (x, None), self.vertices.keys()))
        lowlink = dict(map(lambda x: (x, None), self.vertices.keys()))
        onstack = dict(map(lambda x: (x, False), self.vertices.keys()))
        ret_scc = defaultdict(list)

        def _strongly_connect(v):
            nonlocal index
            indices[v] = index
            lowlink[v] = index
            index = index + 1
            stack.append(v)
            onstack[v] = True

            for w in self.edges[v]:  # edge values not considered here
                if indices[w] is None:
                    _strongly_connect(w)
                    lowlink[v] = min(lowlink[v], lowlink[w])
                elif onstack[w]:
                    lowlink[v] = min(lowlink[v], indices[w])

            if lowlink[v] == indices[v]:
                # v is root vertex of certain scc
                while True:
                    w = stack.pop()
                    onstack[w] = False
                    ret_scc[v].append(w)
                    if w == v:
                        break

        for v in self.vertices:
            if indices[v] is None:
                _strongly_connect(v)

        return ret_scc

    @cached
    def cycles(self) -> List[GraphComponent]:
        return [
            compo
            for _, compo in self.tarjan_scc().items()
            if len(compo) > 1 or compo[0] in self.edges[compo[0]]
        ]

    def zero_indegrees(self) -> List[T]:
        return [v for v in self.vertices if self.indegrees[v] == 0]

    def __len__(self) -> int:
        return len(self.vertices)

    def __copy__(self) -> Self:
        return self.__class__(
            vertices=copy(self.vertices),
            edges={src: copy(dst) for src, dst in self.edges.items()},
            indegrees=copy(self.indegrees),
            outdegrees=copy(self.outdegrees),
        )
