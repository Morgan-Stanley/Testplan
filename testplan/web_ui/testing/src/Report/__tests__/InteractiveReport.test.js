/* Tests the InteractiveReport component. */
import React from 'react';
import {shallow} from 'enzyme';

import InteractiveReport from '../InteractiveReport.js';
import {FakeInteractiveReport} from '../../Common/sampleReports.js';
import {ReportToNavEntry} from '../../Common/utils.js';

describe('InteractiveReport', () => {
  it("shallow renders and matches snapshot", () => {
    const interactiveReport = shallow(<InteractiveReport />);
    expect(interactiveReport).toMatchSnapshot();
  });

  it("updates testcase status to passed", () => {
    const interactiveReport = shallow(<InteractiveReport />);
    interactiveReport.setState({report: FakeInteractiveReport});

    const selectedReportEntries = [
      FakeInteractiveReport,
      FakeInteractiveReport.entries[0],
      FakeInteractiveReport.entries[0].entries[0],
      FakeInteractiveReport.entries[0].entries[0].entries[0],
    ];

    const selectedNavEntries = selectedReportEntries.map(ReportToNavEntry);
    interactiveReport.instance().setEntryStatus(selectedNavEntries, "passed");
    interactiveReport.update();

    const newTestcaseEntry = (
      interactiveReport.state("report").entries[0].entries[0].entries[0]
    );
    expect(newTestcaseEntry.status).toBe("passed");
    expect(interactiveReport).toMatchSnapshot();
  });
});

