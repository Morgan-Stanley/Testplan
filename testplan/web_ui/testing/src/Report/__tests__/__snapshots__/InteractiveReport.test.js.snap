// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InteractiveReport shallow renders and matches snapshot 1`] = `
<div
  className=""
>
  <Toolbar
    handleNavFilter={[Function]}
    status={null}
    updateEmptyDisplayFunc={[Function]}
    updateFilterFunc={[Function]}
    updateTagsDisplayFunc={[Function]}
  />
  <InteractiveNav
    displayEmpty={true}
    displayTags={false}
    filter={null}
    handleNavClick={[Function]}
    handlePlayClick={[Function]}
    report={null}
    selected={Array []}
  />
  <Message
    message="Fetching Testplan report..."
  />
</div>
`;

exports[`InteractiveReport updates testcase status to passed 1`] = `
<div
  className=""
>
  <Toolbar
    handleNavFilter={[Function]}
    status="ready"
    updateEmptyDisplayFunc={[Function]}
    updateFilterFunc={[Function]}
    updateTagsDisplayFunc={[Function]}
  />
  <InteractiveNav
    displayEmpty={true}
    displayTags={false}
    filter={null}
    handleNavClick={[Function]}
    handlePlayClick={[Function]}
    report={
      Object {
        "case_count": Object {
          "failed": 0,
          "passed": 0,
        },
        "entries": Array [
          Object {
            "case_count": Object {
              "failed": 0,
              "passed": 0,
            },
            "category": "multitest",
            "description": null,
            "entries": Array [
              Object {
                "case_count": Object {
                  "failed": 0,
                  "passed": 0,
                },
                "category": "suite",
                "description": null,
                "entries": Array [
                  Object {
                    "case_count": Object {
                      "failed": 0,
                      "passed": 0,
                    },
                    "description": null,
                    "entries": Array [],
                    "logs": Array [],
                    "name": "test_interactive",
                    "name_type_index": Set {},
                    "status": "passed",
                    "status_override": null,
                    "suite_related": false,
                    "tags": Object {},
                    "tags_index": Object {},
                    "timer": Object {},
                    "type": "TestCaseReport",
                    "uid": "ddd",
                  },
                ],
                "logs": Array [],
                "name": "Interactive Suite",
                "name_type_index": Set {},
                "part": null,
                "status": "ready",
                "status_override": null,
                "tags": Object {},
                "tags_index": Object {},
                "timer": Object {},
                "type": "TestGroupReport",
                "uid": "ccc",
              },
            ],
            "logs": Array [],
            "name": "Interactive MTest",
            "name_type_index": Set {},
            "part": null,
            "status": "ready",
            "status_override": null,
            "tags": Object {},
            "tags_index": Object {},
            "timer": Object {},
            "type": "TestGroupReport",
            "uid": "bbb",
          },
        ],
        "meta": Object {},
        "name": "Fake Interactive Report",
        "name_type_index": Set {},
        "status": "ready",
        "status_override": null,
        "tags_index": Object {},
        "timer": null,
        "uid": "aaa",
      }
    }
    selected={Array []}
  />
  <Message
    message="Please select a testcase."
  />
</div>
`;
